src/Hosting/Hosting/src/ActivityFeature.cs

```diff
@@ -0,0 +1,52 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http.Features;
+
+namespace Microsoft.AspNetCore.Hosting
+{
+    /// <summary>
+    /// Default implementation for <see cref="IActivityFeature"/>.
+    /// </summary>
+    internal class ActivityFeature : IActivityFeature
+    {
+        private const int _maxActivityDepth = 5;
+        private Activity? _cachedActivity = null;
+        private bool _activityRetrieved;
+        /// <inheritdoc />
+        public Activity? Activity
+        {
+            get
+            {
+                if (!_activityRetrieved && _cachedActivity is null)
```

Comment:
This could be simplified by explicitly setting the activity from the hostContext. Would have perf benefits of avoiding the Activity.Current async local property.

Reply:
This is exactly what I was thinking.